File: src/ts/magma/app/compile/node/MapNode.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { None } from "../../../../magma/api/option/None";
export class MapNode implements Node {
	type: Option<string>;
	strings: Map<string, string>;
	nodes: Map<string, Node>;
	constructor (type: Option<string>, strings: Map<string, string>, nodes: Map<string, Node>) {
		this.type = type;
		this.strings = strings;
		this.nodes = nodes;
	}
	constructor (type: string) {
		this(new Some<>(type), new HashMap<>(), new HashMap<>())/*unknown*/;
	}
	constructor () {
		this(new None<>(), new HashMap<>(), new HashMap<>())/*unknown*/;
	}
	is(type: string): boolean {
		return this.type.filter((inner: string) => inner.equals(type)/*unknown*/).isPresent()/*unknown*/;
	}
	findNode(key: string): Option<Node> {
		if (this.nodes.containsKey(key)/*unknown*/){
			return new Some<>(this.nodes.get(key))/*unknown*/;
		}
		else {
			return new None<>()/*unknown*/;
		}
	}
	findString(key: string): Option<string> {
		if (this.strings.containsKey(key)/*unknown*/){
			return new Some<>(this.strings.get(key))/*unknown*/;
		}
		else {
			return new None<>()/*unknown*/;
		}
	}
	withNode(key: string, value: Node): MapNode {
		this.nodes.put(key, value)/*unknown*/;
		return this/*unknown*/;
	}
	withString(key: string, value: string): MapNode {
		this.strings.put(key, value)/*unknown*/;
		return this/*unknown*/;
	}
}

File: src/java/magma/app/compile/MapNode.java
package magma.app.compile;

import magma.api.collect.iter.Collector;
import magma.api.collect.iter.Iterable;
import magma.api.collect.list.JavaList;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public final class MapNode implements Node {
    private final Optional<String> maybeType;
    private final Map<String, String> strings;
    private final Map<String, Iterable<Node>> nodeLists;

    public MapNode() {
        this(Optional.empty(), new HashMap<>(), new HashMap<>());
    }

    public MapNode(Optional<String> maybeType, Map<String, String> strings, Map<String, Iterable<Node>> nodeLists) {
        this.maybeType = maybeType;
        this.strings = strings;
        this.nodeLists = nodeLists;
    }

    @Override
    public Node withString(String key, String value) {
        this.strings.put(key, value);
        return this;
    }

    @Override
    public Optional<String> findString(String key) {
        if (this.strings.containsKey(key))
            return Optional.of(this.strings.get(key));
        else
            return Optional.empty();
    }

    @Override
    public Node merge(Node other) {
        return other.collect(new StringNodeCollector(this));
    }

    @Override
    public Node withNodeList(String key, Iterable<Node> values) {
        this.nodeLists.put(key, values);
        return this;
    }

    @Override
    public Optional<Iterable<Node>> findNodeList(String key) {
        if (this.nodeLists.containsKey(key))
            return Optional.of(this.nodeLists.get(key));
        else
            return Optional.empty();
    }

    @Override
    public String display() {
        return this.strings.toString() + this.nodeLists.toString();
    }

    @Override
    public boolean is(String type) {
        return this.maybeType.isPresent() && this.maybeType.get()
                .equals(type);
    }

    @Override
    public Node retype(String type) {
        return new MapNode(Optional.of(type), this.strings, this.nodeLists);
    }

    @Override
    public Node collect(Collector<Map.Entry<String, String>, Node> collector) {
        return new JavaList<>(new ArrayList<>(this.strings.entrySet())).iter()
                .collect(collector);
    }
}